File: backend/scripts/gen_questions.js
// 题库 SQL 生成器 - Node 版
const fs = require('fs');

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function pick(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

const CATS = [
  { id: 1, name: '计算机基础', tag: 'CS' },
  { id: 2, name: '前端开发', tag: 'FE' },
  { id: 3, name: '后端开发', tag: 'BE' },
  { id: 4, name: '算法与数据结构', tag: 'ALGO' },
  { id: 5, name: '数据库', tag: 'DB' },
  { id: 6, name: '网络与安全', tag: 'NET' },
];

const SINGLE_TITLES = [
  '关于 {S} 的描述，下列哪项是正确的？',
  '以下哪个选项最准确地描述了 {S} 的特性？',
  '在 {S} 领域中，下列说法正确的是？',
  '{S} 中常用的概念是以下哪一个？',
  '下列关于 {S} 的说法中，正确的是？',
  '下列哪个不属于 {S}？',
  '下列哪个是 {S} 领域的核心概念？',
  '关于 {S} 原理，下列哪项陈述最准确？',
];

const SINGLE_OPTIONS = [
  '支持事务 ACID 特性', '提供强类型检查', '运行在 JVM 上', '采用单线程模型',
  '基于事件驱动', '采用协程机制', '支持反射', '支持泛型', '拥有闭包特性',
  '提供依赖注入', '采用 MVC 架构', '支持热加载', '采用单例模式',
  '支持懒加载', '采用工厂模式', '使用观察者模式', '支持装饰器',
  '提供中间件机制', '采用 ORM 映射', '支持 SQL 注入防护', '采用连接池',
  '使用负载均衡', '采用微服务架构', '支持容器化部署', '使用分布式锁',
];

function escape(str) {
  return str.replace(/'/g, "''").replace(/\\/g, '\\\\');
}

function jsonStr(s) {
  return JSON.stringify(s).replace(/'/g, "''");
}

function singleQ(catName, idx, seed) {
  const tmpl = pick(SINGLE_TITLES);
  const title = tmpl.replace('{S}', catName);
  const correctIdx = randInt(0, 3);
  const usedOpts = new Set();
  const opts = [];
  const keys = ['A', 'B', 'C', 'D'];
  for (let i = 0; i < 4; i++) {
    let opt;
    let tries = 0;
    do {
      opt = pick(SINGLE_OPTIONS) + '（' + catName + '相关）';
      tries++;
    } while (usedOpts.has(opt) && tries < 10);
    usedOpts.add(opt);
    if (i === correctIdx) {
      opt = catName + '的核心特征之一（' + (seed % 1000) + ')';
    }
    opts.push({ key: keys[i], text: opt });
  }
  return {
    title,
    options: opts,
    answer: [keys[correctIdx]],
    analysis: '解析：本题考查' + catName + '相关知识点。正确答案' + keys[correctIdx] + '。',
  };
}

function multiQ(catName, idx, seed) {
  const correctCount = randInt(2, 3);
  const allIdx = [0, 1, 2, 3];
  // shuffle
  for (let i = allIdx.length - 1; i > 0; i--) {
    const j = randInt(0, i);
    [allIdx[i], allIdx[j]] = [allIdx[j], allIdx[i]];
  }
  const correctIdx = allIdx.slice(0, correctCount).sort((a, b) => a - b);
  const keys = ['A', 'B', 'C', 'D'];
  const opts = [];
  const correctSet = new Set(correctIdx);
  for (let i = 0; i < 4; i++) {
    let text;
    if (correctSet.has(i)) {
      text = catName + '的正确特性之一（' + (seed % 100) + '）';
    } else {
      text = '干扰项：' + pick(SINGLE_OPTIONS);
    }
    opts.push({ key: keys[i], text });
  }
  return {
    title: '以下哪些属于 ' + catName + ' 的相关特性？（多选）',
    options: opts,
    answer: correctIdx.map(i => keys[i]),
    analysis: '解析：本题为多选题，正确答案为 ' + correctIdx.map(i => keys[i]).join(',') + '。',
  };
}

function judgeQ(catName, idx, seed) {
  const isTrue = randInt(0, 1) === 1;
  const trueStatements = [
    'TCP 是面向连接的可靠传输协议。',
    'HTTP 协议默认使用 80 端口。',
    '主键不允许重复，也不允许为空。',
    'CSS 中 margin 控制元素外边距。',
    'InnoDB 存储引擎支持事务。',
    'JavaScript 是单线程语言。',
    'Vue 3 引入组合式 API。',
    'Go 语言支持 goroutine 并发。',
    'MySQL 使用 SQL 作为查询语言。',
    'HTTPS 默认端口是 443。',
  ];
  const falseStatements = [
    'Java 是弱类型语言。',
    'HTTP 是面向连接的状态协议。',
    'MyISAM 支持事务回滚。',
    '主键允许为空。',
    'TCP 三次握手是四次挥手。',
    'Go 不支持并发。',
    'SQL 中 DELETE 用于查询。',
    'HTTPS 比 HTTP 更不安全。',
    '数据库索引越多查询越快。',
    'IPv4 地址用 8 位表示。',
  ];
  const stmt = isTrue ? pick(trueStatements) : pick(falseStatements);
  return {
    title: '【判断】' + stmt,
    options: null,
    answer: [isTrue ? 'true' : 'false'],
    analysis: '解析：正确答案是' + (isTrue ? '正确' : '错误') + '。',
  };
}

function fillQ(catName, idx, seed) {
  const keywords = ['algorithm', 'function', 'variable', 'class', 'interface', 'module', 'package', 'method', 'array', 'pointer', 'coroutine', 'mutex', 'channel', 'struct', 'enum', 'lambda', 'closure', 'recursion', 'iterator', 'decorator', 'middleware', 'router', 'controller', 'service', 'repository', 'entity', 'dto', 'vo', 'bo', 'dao', 'po', 'orm', 'jwt', 'session', 'cookie', 'cache', 'queue', 'stack', 'heap', 'tree', 'graph', 'hash', 'binary', 'index', 'trigger', 'view', 'cursor', 'transaction', 'isolation', 'rollback', 'commit', 'lock', 'deadlock', 'sql', 'nosql', 'acid', 'cap', 'base'];
  const kw = pick(keywords);
  return {
    title: '【填空】在计算机相关领域中，关键术语英文是 ____（请填入英文单词）。',
    options: null,
    answer: [kw],
    analysis: '解析：答案为 ' + kw + '。',
  };
}

function essayQ(catName, idx, seed) {
  const topics = [
    '请简述相关概念及其典型应用场景。',
    '对比分析两种方案的优缺点及适用场景。',
    '描述其工作原理，并举例说明。',
    '分析常见问题及解决方案。',
    '总结最佳实践和注意事项。',
    '谈谈性能优化的常用手段。',
    '说明安全方面的风险和防护措施。',
    '从架构演进角度分析其发展。',
  ];
  return {
    title: '【简答】' + pick(topics),
    options: null,
    answer: ['open'],
    analysis: '主观题，由教师人工批改。',
  };
}

function codeQ(catName, idx, seed) {
  const problems = [
    '实现一个函数，输入数组，输出最大值。',
    '判断字符串是否为回文。',
    '实现二分查找算法。',
    '合并两个有序数组。',
    '实现快速排序。',
    '实现 LRU 缓存。',
    '实现单例模式。',
    '反转单链表。',
    '实现栈的数据结构。',
    '实现二叉树的前序遍历。',
    '实现两数之和（哈希表法）。',
    '求数组的最大子序和。',
    '判断链表是否有环。',
    '实现冒泡排序。',
    '实现选择排序。',
    '实现插入排序。',
    '字符串转整数 (atoi)。',
    '最长公共前缀。',
    '有效的括号匹配。',
    '爬楼梯问题。',
  ];
  return {
    title: '【编程】' + pick(problems),
    options: null,
    answer: ['code'],
    analysis: '编程题，需提交代码评测。',
  };
}

// ============ 生成 SQL ============
const ROWS = [];
let idCounter = 107; // 现有最大 106

function addRow(cat, type, q, tag, score, difficulty) {
  ROWS.push({
    id: idCounter++,
    category_id: cat.id,
    type,
    difficulty,
    title: q.title,
    options: q.options,
    answer: q.answer,
    analysis: q.analysis,
    tag,
    score,
    creator_id: 1,
    status: 1,
  });
}

// 1. 单选（每类约 80 题，共 ~500）
for (const cat of CATS) {
  for (let i = 0; i < 85; i++) {
    const q = singleQ(cat.name, i, idCounter);
    addRow(cat, 1, q, cat.tag, 2, randInt(1, 3));
  }
}

// 2. 多选（每类约 35 题，共 ~200）
for (const cat of CATS) {
  for (let i = 0; i < 35; i++) {
    const q = multiQ(cat.name, i, idCounter);
    addRow(cat, 2, q, cat.tag, 3, randInt(1, 3));
  }
}

// 3. 判断（每类约 20-40，共 150）
for (const cat of CATS) {
  for (let i = 0; i < 25; i++) {
    const q = judgeQ(cat.name, i, idCounter);
    addRow(cat, 3, q, '判断', 2, randInt(1, 2));
  }
}

// 4. 填空（每类 25-40，共 150）
for (const cat of CATS) {
  for (let i = 0; i < 30; i++) {
    const q = fillQ(cat.name, i, idCounter);
    addRow(cat, 4, q, '填空', 3, randInt(1, 3));
  }
}

// 5. 简答（每类 15，共 90）
for (const cat of CATS) {
  for (let i = 0; i < 15; i++) {
    const q = essayQ(cat.name, i, idCounter);
    addRow(cat, 5, q, '简答', 8, randInt(2, 3));
  }
}

// 6. 编程（80）
for (let i = 0; i < 50; i++) {
  const q = codeQ('后端', i, idCounter);
  addRow(CATS[2], 6, q, '编程', 15, 3);
}
for (let i = 0; i < 30; i++) {
  const q = codeQ('算法', i, idCounter);
  addRow(CATS[3], 6, q, '编程', 15, 3);
}

console.log('生成题目数: ' + ROWS.length);
console.log('ID 范围: ' + ROWS[0].id + ' - ' + ROWS[ROWS.length - 1].id);

// ============ 输出 SQL ============
const HEAD = `-- 自动生成的 1000+ 题库种子数据
-- 生成时间: ${new Date().toISOString()}
-- 共生成 ${ROWS.length} 题（不含原有 106 题）

INSERT INTO ke_question (id, category_id, type, difficulty, title, options, answer, analysis, tags, score, creator_id, status) VALUES
`;

const values = ROWS.map(r => {
  const options = r.options ? `'[${r.options.map(o => `{"key":"${o.key}","text":"${escape(o.text)}"}`).join(',')}]'` : 'NULL';
  const answer = `'[${r.answer.map(a => `"${escape(a)}"`).join(',')}]'`;
  const title = `'${escape(r.title)}'`;
  const analysis = `'${escape(r.analysis)}'`;
  const tag = `'${escape(r.tag)}'`;
  return `  (${r.id}, ${r.category_id}, ${r.type}, ${r.difficulty}, ${title}, ${options}, ${answer}, ${analysis}, ${tag}, ${r.score}, ${r.creator_id}, ${r.status})`;
}).join(',\n');

const SQL = HEAD + values + ';\n';

fs.writeFileSync(process.argv[2] || 'gen_questions.sql', SQL, 'utf-8');
console.log('SQL 已写入: ' + (process.argv[2] || 'gen_questions.sql'));
console.log('SQL 大小: ' + (SQL.length / 1024).toFixed(1) + ' KB');

File: backend/scripts/gen_questions.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"

	_ "github.com/go-sql-driver/mysql"
)

// 题目模板库
type Template struct {
	CategoryID int64
	Type       int // 1-6
	Title      string
	Options    []map[string]string
	Answer     []string
	Analysis   string
	Tag        string
	Score      int
}

var singleSubjects = []string{
	"计算机基础", "网络协议", "操作系统", "数据结构", "编程语言",
	"前端框架", "后端开发", "数据库", "算法", "安全",
}

func main() {
	rand.Seed(time.Now().UnixNano())
	db, err := sql.Open("mysql", "root:123456@tcp(127.0.0.1:3306)/KoalaExam?charset=utf8mb4")
	if err != nil { fmt.Println("open:", err); os.Exit(1) }
	defer db.Close()

	var maxID int64
	db.QueryRow("SELECT COALESCE(MAX(id), 0) FROM ke_question").Scan(&maxID)
	fmt.Printf("现有最大 id: %d\n", maxID)

	var totalBefore int64
	db.QueryRow("SELECT COUNT(*) FROM ke_question").Scan(&totalBefore)
	fmt.Printf("现有题目数: %d\n", totalBefore)

	// 生成 1100 道题（覆盖各类）
	templates := buildTemplates()
	fmt.Printf("生成模板数: %d\n", len(templates))

	// 批量插入
	tx, err := db.Begin()
	if err != nil { fmt.Println("tx:", err); os.Exit(1) }

	stmt, err := tx.Prepare(`INSERT INTO ke_question
		(id, category_id, type, difficulty, title, options, answer, analysis, tags, score, creator_id, status)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`)
	if err != nil { fmt.Println("prepare:", err); tx.Rollback(); os.Exit(1) }
	defer stmt.Close()

	inserted := int64(0)
	for i, t := range templates {
		id := maxID + int64(i+1)
		categoryID := t.CategoryID
		diff := rand.Intn(3) + 1 // 1-3
		creator := int64(1)
		status := 1

		var optionsStr, answerStr sql.NullString
		if t.Options != nil {
			b, _ := json.Marshal(t.Options)
			optionsStr = sql.NullString{String: string(b), Valid: true}
		}
		if t.Answer != nil {
			b, _ := json.Marshal(t.Answer)
			answerStr = sql.NullString{String: string(b), Valid: true}
		}

		_, err := stmt.Exec(id, categoryID, t.Type, diff, t.Title, optionsStr, answerStr, t.Analysis, t.Tag, t.Score, creator, status)
		if err != nil {
			fmt.Printf("insert id=%d err: %v\n", id, err)
			continue
		}
		inserted++
	}

	if err := tx.Commit(); err != nil { fmt.Println("commit:", err); os.Exit(1) }

	fmt.Printf("✅ 成功插入 %d 道题\n", inserted)

	var totalAfter int64
	db.QueryRow("SELECT COUNT(*) FROM ke_question").Scan(&totalAfter)
	fmt.Printf("插入后总数: %d\n", totalAfter)

	// 按类型统计
	rows, _ := db.Query("SELECT type, COUNT(*) FROM ke_question GROUP BY type ORDER BY type")
	defer rows.Close()
	fmt.Println("\n按类型统计:")
	for rows.Next() {
		var t int
		var c int64
		rows.Scan(&t, &c)
		fmt.Printf("  type=%d (%s): %d\n", t, typeName(t), c)
	}

	rows2, _ := db.Query("SELECT category_id, COUNT(*) FROM ke_question GROUP BY category_id ORDER BY category_id")
	defer rows2.Close()
	fmt.Println("\n按分类统计:")
	for rows2.Next() {
		var c int64
		var cnt int64
		rows2.Scan(&c, &cnt)
		fmt.Printf("  cat=%d: %d\n", c, cnt)
	}
}

func typeName(t int) string {
	switch t {
	case 1: return "单选"
	case 2: return "多选"
	case 3: return "判断"
	case 4: return "填空"
	case 5: return "简答"
	case 6: return "编程"
	}
	return "未知"
}

func buildTemplates() []Template {
	var ts []Template

	// ===== 单选题 (type=1) - 500道 =====
	ts = append(ts, generateSingle(1, "计算机基础", "CS", 80)...)
	ts = append(ts, generateSingle(2, "前端开发", "FE", 80)...)
	ts = append(ts, generateSingle(3, "后端开发", "BE", 80)...)
	ts = append(ts, generateSingle(4, "算法与数据结构", "ALGO", 80)...)
	ts = append(ts, generateSingle(5, "数据库", "DB", 90)...)
	ts = append(ts, generateSingle(6, "网络与安全", "NET", 90)...)
	_ = singleSubjects

	// ===== 多选题 (type=2) - 200道 =====
	ts = append(ts, generateMulti(1, "计算机基础", "CS", 35)...)
	ts = append(ts, generateMulti(2, "前端开发", "FE", 35)...)
	ts = append(ts, generateMulti(3, "后端开发", "BE", 35)...)
	ts = append(ts, generateMulti(5, "数据库", "DB", 45)...)
	ts = append(ts, generateMulti(6, "网络与安全", "NET", 50)...)

	// ===== 判断题 (type=3) - 150道 =====
	ts = append(ts, generateJudge(1, 25)...)
	ts = append(ts, generateJudge(2, 25)...)
	ts = append(ts, generateJudge(3, 25)...)
	ts = append(ts, generateJudge(5, 35)...)
	ts = append(ts, generateJudge(6, 40)...)

	// ===== 填空题 (type=4) - 150道 =====
	ts = append(ts, generateFill(1, 30)...)
	ts = append(ts, generateFill(3, 35)...)
	ts = append(ts, generateFill(5, 40)...)
	ts = append(ts, generateFill(2, 25)...)
	ts = append(ts, generateFill(6, 20)...)

	// ===== 简答题 (type=5) - 80道 =====
	ts = append(ts, generateEssay(1, 15)...)
	ts = append(ts, generateEssay(3, 20)...)
	ts = append(ts, generateEssay(4, 15)...)
	ts = append(ts, generateEssay(5, 15)...)
	ts = append(ts, generateEssay(2, 15)...)

	// ===== 编程题 (type=6) - 80道 =====
	ts = append(ts, generateCode(3, 25)...)
	ts = append(ts, generateCode(4, 35)...)
	ts = append(ts, generateCode(2, 20)...)

	return ts
}

// ============ 单选 ============
func generateSingle(cat int64, name, tag string, n int) []Template {
	rand.Seed(time.Now().UnixNano() + int64(n*cat))
	out := make([]Template, 0, n)
	for i := 0; i < n; i++ {
		opts, ans := randomSingleQA(name, i+1)
		out = append(out, Template{
			CategoryID: cat,
			Type:       1,
			Title:      fmt.Sprintf("[%s-单选-%04d] %s", name, i+1, singleTitles(name, i)),
			Options:    opts,
			Answer:     []string{ans},
			Analysis:   fmt.Sprintf("解析：本题考查 %s 知识点。正确答案 %s。", name, ans),
			Tag:        tag,
			Score:      2,
		})
	}
	return out
}

func singleTitles(name string, i int) string {
	templates := []string{
		fmt.Sprintf("关于 %s 的描述，下列哪项是正确的？", name),
		fmt.Sprintf("以下哪个选项最准确地描述了 %s 的特性？", name),
		fmt.Sprintf("在 %s 领域中，下列说法正确的是？", name),
		fmt.Sprintf("%s 中常用的概念是以下哪一个？", name),
		fmt.Sprintf("下列关于 %s 的说法中，正确的是？", name),
	}
	return templates[i%len(templates)]
}

func randomSingleQA(name string, seed int) ([]map[string]string, string) {
	correctIdx := rand.Intn(4)
	keys := []string{"A", "B", "C", "D"}
	corrects = []string{}
	for i, k := range keys {
		text := fmt.Sprintf("选项 %s：%s 相关描述 %d", k, name, seed+i)
		if i == correctIdx {
			text = fmt.Sprintf("选项 %s：%s 的正确描述 %d", k, name, seed)
		}
		opts = append(opts, map[string]string{"key": k, "text": text})
	}
	return opts, keys[correctIdx]
}

// ============ 多选 ============
func generateMulti(cat int64, name, tag string, n int) []Template {
	out := make([]Template, 0, n)
	for i := 0; i < n; i++ {
		opts, ans := randomMultiQA(name, i+1)
		out = append(out, Template{
			CategoryID: cat,
			Type:       2,
			Title:      fmt.Sprintf("[%s-多选-%04d] 以下哪些是 %s 的相关特性？（多选）", name, i+1, name),
			Options:    opts,
			Answer:     ans,
			Analysis:   fmt.Sprintf("解析：本题为多选题，答案为 %s。", strings.Join(ans, ",")),
			Tag:        tag,
			Score:      3,
		})
	}
	return out
}

func randomMultiQA(name string, seed int) ([]map[string]string, []string) {
	correctCount := 2 + rand.Intn(2) // 2-3 个正确答案
	selected := rand.Perm(4)[:correctCount]
	keySet := map[string]bool{}
	ans := []string{}
	for _, idx := range selected {
		keySet[[]"A","B","C","D"[idx]] = true
		ans = append(ans, []string{"A","B","C","D"}[idx])
	}
	keys := []string{"A", "B", "C", "D"}
	opts := make([]map[string]string, 0, 4)
	for i, k := range keys {
		text := fmt.Sprintf("特性 %s：%s 知识点 %d", k, name, seed+i)
		if keySet[k] {
			text = fmt.Sprintf("正确特性 %s：%s 描述 %d", k, name, seed)
		}
		opts = append(opts, map[string]string{"key": k, "text": text})
	}
	return opts, ans
}

// ============ 判断 ============
func generateJudge(cat int64, n int) []Template {
	out := make([]Template, 0, n)
	for i := 0; i < n; i++ {
		isTrue := rand.Intn(2) == 0
		ans := "false"
		if isTrue { ans = "true" }
		out = append(out, Template{
			CategoryID: cat,
			Type:       3,
			Title:      fmt.Sprintf("[判断-%04d] 这是关于计算机知识的一条陈述，请判断对错：相关描述 %d-%d。", i+1, cat, i),
			Answer:     []string{ans},
			Analysis:   fmt.Sprintf("解析：正确答案是 %s。", map[string]string{"true":"正确","false":"错误"}[ans]),
			Tag:        "判断",
			Score:      2,
		})
	}
	return out
}

// ============ 填空 ============
func generateFill(cat int64, n int) []Template {
	out := make([]Template, 0, n)
	for i := 0; i < n; i++ {
		keywords := []string{"algorithm", "function", "variable", "class", "interface", "module", "package", "method"}
		kw := keywords[rand.Intn(len(keywords))]
		out = append(out, Template{
			CategoryID: cat,
			Type:       4,
			Title:      fmt.Sprintf("[填空-%04d] 在计算机科学中，相关的关键术语是 ____ （请填入英文术语）。", i+1),
			Answer:     []string{kw},
			Analysis:   fmt.Sprintf("解析：答案为 %s。", kw),
			Tag:        "填空",
			Score:      3,
		})
	}
	return out
}

// ============ 简答 ============
func generateEssay(cat int64, n int) []Template {
	out := make([]Template, 0, n)
	topics := []string{
		"请简述相关概念及其应用场景。",
		"对比分析两种方案的优缺点。",
		"描述其工作原理，并举例说明。",
		"分析常见问题及解决方案。",
		"总结最佳实践。",
	}
	for i := 0; i < n; i++ {
		out = append(out, Template{
			CategoryID: cat,
			Type:       5,
			Title:      fmt.Sprintf("[简答-%04d] %s", i+1, topics[i%len(topics)]),
			Answer:     []string{"open"},
			Analysis:   "主观题，需人工批改。",
			Tag:        "简答",
			Score:      8,
		})
	}
	return out
}

// ============ 编程 ============
func generateCode(cat int64, n int) []Template {
	out := make([]Template, 0, n)
	problems := []string{
		"实现一个函数，输入数组，输出最大值。",
		"判断字符串是否为回文。",
		"实现二分查找。",
		"实现两个有序数组合并。",
		"实现快速排序。",
		"实现 LRU 缓存。",
		"实现单例模式。",
		"反转链表。",
	}
	for i := 0; i < n; i++ {
		out = append(out, Template{
			CategoryID: cat,
			Type:       6,
			Title:      fmt.Sprintf("[编程-%04d] %s", i+1, problems[i%len(problems)]),
			Answer:     []string{"code"},
			Analysis:   "编程题，需提交代码并自动/人工评测。",
			Tag:        "编程",
			Score:      15,
		})
	}
	return out
}
